Source/levels/themes.cpp

```diff
@@ -97,51 +83,37 @@ bool TFit_Shrine(int i)
 	return true;
 }
 
-bool TFit_Obj5(int t)
+bool CheckThemeObj5(Point origin, int8_t regionId)
 {
-	int xp = 0;
-	int yp = 0;
-	int r = GenerateRnd(5) + 1;
-	int rs = r;
-
-	while (r > 0) {
-		bool found = false;
-		if (dTransVal[xp][yp] == themes[t].ttval && IsTileNotSolid({ xp, yp })) {
-			found = true;
-			for (int i = 0; found && i < 25; i++) {
-				if (TileHasAny(dPiece[xp + trm5x[i]][yp + trm5y[i]], TileProperties::Solid)) {
-					found = false;
-				}
-				if (dTransVal[xp + trm5x[i]][yp + trm5y[i]] != themes[t].ttval) {
-					found = false;
-				}
-			}
+	const PointsInRectangleRange searchArea { Rectangle { origin, 2 } };
+	return std::all_of(searchArea.cbegin(), searchArea.cend(), [regionId](Point testPosition) {
+		// note out-of-bounds tiles are not solid, this function relies on the guard in TFit_Obj5 and dungeon border
+		if (IsTileSolid(testPosition)) {
+			return false;
+		}
+		// If the theme object would extend into a different region then it doesn't fit.
+		if (dTransVal[testPosition.x][testPosition.y] != regionId) {
+			return false;
 		}
+		return true;
+	});
+}
 
-		if (!found) {
-			xp++;
-			if (xp == MAXDUNX) {
-				xp = 0;
-				yp++;
-				if (yp == MAXDUNY) {
-					if (r == rs) {
-						return false;
-					}
-					yp = 0;
-				}
+bool TFit_Obj5(int t)
+{
+	AdvanceRndSeed();

```

Comment:
Hmm, this seems to actually be discarding the first x number of potential spots, I'll need to restore that.